lib/discord_bot/discord_bot.ex

```diff
@@ -0,0 +1,188 @@
+defmodule DiscordSupervisor do
+  use Supervisor
+  alias Alchemy.Client
+  alias Alchemy.Embed
+  require Logger
+
+  defmodule DiscordBot do
+    use GenServer
+
+    defmodule HubsCommands do
+      use Alchemy.Cogs
+      require Alchemy.Embed, as: Embed
+
+      Cogs.def duck do
+        Cogs.say("Quack :duck:")
+      end
+
+      Cogs.def bound do
+        case Cachex.get!(:discord_bot_state, "hubs_for_channel")[message.channel_id] do
+          hub_ids = [_ | _] ->
+            Cogs.say("This channel is bound to #{Enum.join(hub_ids, ", ")}")
+
+          _ ->
+            Cogs.say("No hubs bound to this channel")
+        end
+      end
+    end
+
+    defmodule HubsEvents do
+      use Alchemy.Events
+
+      Events.on_channel_update(:on_channel_update)
+      Events.on_ready(:on_ready)
+      Events.on_message(:on_message)
+
+      def on_channel_update(_channel) do
+        update_bound_channels()
+      end
+
+      def on_ready(_shard, _total_shards) do
+        update_bound_channels()
+      end
+
+      def on_message(msg) do
+        if hub_ids = !msg.author.bot && Cachex.get!(:discord_bot_state, "hubs_for_channel")[msg.channel_id] do
+          Logger.debug "Outgoing message from Discord to Hubs: #{msg.content}"
+          hub_ids |> Enum.each(&broadcast_to_hubs(&1, msg))
+        end
+      end
+
+      def broadcast_to_hubs(hub_id, msg) do
+        RetWeb.Endpoint.broadcast!("hub:#{hub_id}", "message", %{
+              type: "chat",
+              body: msg.content,
+              from: "#{msg.author.username}##{msg.author.discriminator}"
+                                   })
+      end
+
+      def update_bound_channels do
+        hostnames = Application.get_env(:ret, Elixir.DiscordBot)[:hostnames] |> String.split
+        host_clauses = hostnames |> Enum.map(&("#{&1}(?:\\:\\d+)?")) |> Enum.join("|")
+        {:ok, hub_url_regex} = Regex.compile("https?://(?:#{host_clauses})/(\\w+)/?\\S*")
+
+        bound_channels =
+        for guild <- elem(Alchemy.Client.get_current_guilds(), 1),
+          channel <- elem(Alchemy.Client.get_channels(guild.id), 1),
+          topic = channel.topic,
+          matches = Regex.scan(hub_url_regex, topic),
+          [_, hub_id] <- matches do
+          {hub_id, channel.id}
+        end
+
+        channels_by_hub = bound_channels |> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
+        hubs_by_channel = bound_channels |> Enum.group_by(&elem(&1, 1), &elem(&1, 0))
+
+        Logger.debug "Bound Discord channels per hub: #{inspect(channels_by_hub)}"
+        Cachex.transaction!(:discord_bot_state, ["hubs_for_channel", "channels_for_hub"], fn(cache) ->
```

Comment:
i'd suggest using atoms not strings for these keys if possible

Reply:
Agreed.